[FLINK-14189][runtime] Extend TaskExecutor to support dynamic slot allocation

xintongsong opened the pull request:
## What is the purpose of the change

This PR is part fo FLIP-56. It extends `TaskExecutor` to support dynamic slot allocation.
With this PR, the `TaskExecutor` now supports both previous static slot allocation and the new dynamic allocation.
- For dynamic slot allocation, a special `SlotID` (with negative slot index) needs to be provided, as well as a `ResourceProfile` to describe the requested slot resources.
- For static slot allocation, a normal `SlotID` (with slot index in the range of `[0, slotNumber)`) needs to be provided. In such cases, the provided `ResourceProfile` will be ignored, and the default slot resource profile derived from configuration will be used.

This PR is based on #10146.

## Brief change log

- 75d805dfaecb953e72cf84a4439e549a26118539..b083c68650acd6b097dc7049276df080196db26f: Commits of previous PRs.
- aaf47129e5cc2adfd03c0c96749745b22937ee57: Hotfix of refactoring `TaskSlotTable` and `TaskSlot`.
- 0fec80f46457a9f7bd4fdd783c68d4acfd1d65d6: Dynamically create a slot on allocation and destroy a slot on free.
- 0fe0b6618d55bb1d111ed22f7692806bdb07075d: Support allocate slots without pre-defined slot id.
- 8c661dbf866f0153c2c41b370c766e627817a0a6: Introduce `ResourceProfileBudgetManager.` for managing resource reserving and releasing of slots.
- 8d126889bb9986eb575656e2be028c5357e3f1e2: Use `ResourceProfileBudgetManager.` to bookkeep allocated / available resources on the TM side.
- a005c509ee928ea0e30d47a5a94946bd59622166: Generate slot report with TM available resources.
- c54d7b94b9ca46e6b8dc2a072919ab0b1deed022: Task executor supports allocating slots with dynamic resources.

## Verifying this change

Added unit test cases in `TaskSlotTableTest` and `TaskExecutorTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java:
```diff
@@ -69,6 +70,22 @@ public static ResourceProfile createDefaultSlotResourceProfile() {
+			if (numberOfSlots % 2 != 0) {
+				result = result.merge(defaultResourceProfile);
+			}
+			numberOfSlots = numberOfSlots / 2;
```
I would keep it as simple loop merging `defaultResourceProfile` `numberOfSlots - 1` times for simpler readability.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java:
```diff
@@ -242,10 +263,15 @@ public boolean allocateSlot(int index, JobID jobId, AllocationID allocationId, T
 
+		ResourceProfile slotResourceProfile = defaultSlotResourceProfile;
+		if (!resourceProfileBookkeeper.reserve(slotResourceProfile)) {
+			return false;
```
would be nice to log how much we tried to allocate and what was available budget in case of failure

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/SlotReport.java:
```diff
@@ -46,18 +50,33 @@ public SlotReport(SlotStatus slotStatus) {
 	}
 
 	public SlotReport(final Collection<SlotStatus> slotsStatus) {
+		this(slotsStatus,
```
we should annotate the first 2 constructors with `@VisibleForTesting`

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java:
```diff
@@ -854,6 +854,7 @@ private void allocateSlot(TaskManagerSlot taskManagerSlot, PendingSlotRequest pe
 			slotId,
 			pendingSlotRequest.getJobId(),
 			allocationId,
+			ResourceProfile.ZERO,
```
`pendingSlotRequest.getResourceProfile()`

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java:
```diff
@@ -196,16 +217,21 @@ public SlotReport createSlotReport(ResourceID resourceId) {
 		boolean result = taskSlot.allocate(jobId, allocationId);
 
 		if (result) {
+			taskSlots.put(index, taskSlot);
```
True, `result` should always be `true`. The idea here is to make `TaskSlotTable` independent from the implementation of `TaskSlot`, only stick to the its contract (which is that `TaskSlot#allocate` might fail).

Agree that we should do sanity check on the duplicated `index`.

xintongsong replied:
As discussed offline, I'll remove `TaskSlotState#FREE` (task slot will be created with the state `ALLOCATED`), and `TaskSlot#allocate` as well (because task slot can no longer be allocated once created).

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java:
```diff
@@ -412,7 +441,7 @@ public boolean tryMarkSlotActive(JobID jobId, AllocationID allocationId) {
 		TaskSlot taskSlot = taskSlots.get(index);
 
-		return taskSlot.isFree();
+		return taskSlot == null || taskSlot.isFree();
```
True, we could consider removing `TaskSlotState#FREE`, but why `ACTIVE`?

xintongsong replied:
See above. I'll remove `TaskSlotState#FREE` but keep `ACTIVE` as discussed offline. I'll also remove `TaskSlot#isFree` because there should never be a free task slot anymore.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java:
```diff
@@ -205,10 +221,11 @@ public SlotReport createSlotReport(ResourceID resourceId) {
-	 * Allocate the slot with the given index for the given job and allocation id. Returns true if
-	 * the slot could be allocated. Otherwise it returns false.
+	 * Allocate the slot with the given index for the given job and allocation id. If negative index is
+	 * given, a new auto increasing index will be generated. Returns true if the slot could be allocated.
```
I think eventually we are going to remove `SlotID` and use `AllocationID` to identify the slots. But maybe not now, since we still need `SlotID` for the legacy RM.

Since we're not removing the `SlotID` at the moment, I'm in favor of keep the `SlotID` of dynamic slots valid, in other words each slot has a unique `SlotID`. Thus, we need to crate a unique id for each dynamic slot anyway, no matter positive or negative.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfileBookkeeper.java:
```diff
@@ -0,0 +1,75 @@
+ *
+ * <p>Both the total budget and the reservations are in the form of {@link ResourceProfile}.
+ */
+public class ResourceProfileBookkeeper {
```
How about `ResourceProfileBudgetManager`, similar to `KeyedBudgetManager`?

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfileBookkeeper.java:
```diff
@@ -0,0 +1,75 @@
+		return availableBudget;
+	}
+
+	public synchronized boolean reserve(final ResourceProfile reservation) {
```
I was thinking about making this a general class, so it can be reused by TM and later RM, and other potential use cases in the future. Then maybe we should not only consider the scenario of slot table, where it is always accessed from TM main thread.
Besides, slot allocation / free should not be very frequent, so the cost of making this class thread safe should be affordable.

xintongsong replied:
As discussed offline, we keep this class thread unsafe.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java:
```diff
@@ -296,22 +291,11 @@ private static TaskSlotTable createTaskSlotTable(
-		return new TaskSlotTable(createTaskSlotsFromResources(resourceProfiles, pageSize), timerService);
-	}
-
-	private static List<TaskSlot> createTaskSlotsFromResources(
```
As discussed offline, soon in later commits we no longer create the entire list of `TaskSlot` when creating `TaskSlotTable`. So there's no need to move the whole `createTaskSlotsFromResources` because it will soon be removed anyway.

xintongsong commented:
Thanks for the review, @azagrebin. The comments are addressed.

xintongsong commented:
Travis passed: https://travis-ci.org/xintongsong/flink/builds/619623328

xintongsong commented:
Thanks for the review, @azagrebin.
I've updated the PR, addressed your comments, and added a commit for registering TM at RM with total resource profile as we discussed offline.

azagrebin commented:
Thanks for addressing comments @xintongsong
LGTM, merging this
merged into master by da071ce3c26619d2fbd337a165f2405a75c4d91b..db3dec55d545989642510f3bf8f307605e732a6d
